### PR TITLE
Add wallet backup text to import* and add* RPCs

### LINKED ISSUE #11243: add* RPCs require backups
At a minimum, help text should note this. This is true for things like addwitnessaddress especially.

### PR DESCRIPTION
**Pull request by meshcollider**
Closes https://github.com/bitcoin/bitcoin/issues/11243

Adds "Requires a new wallet backup" text to `addwitnessaddress`, `importprivkey`, `importmulti`, `importaddress`, `importpubkey`, and `addmultisigaddress`. Also adds a warning to `dumpwallet` that backing up the seed alone is not sufficient to back up non-HD addresses

### CODE REVIEW

File: src/wallet/rpcdump.cpp [hunk 1]
```diff
@@ -62,7 +62,7 @@ std::string DecodeDumpString(const std::string &str) {
     for (unsigned int pos = 0; pos < str.length(); pos++) {
         unsigned char c = str[pos];
         if (c == '%' && pos+2 < str.length()) {
-            c = (((str[pos+1]>>6)*9+((str[pos+1]-'0')&15)) << 4) | 
+            c = (((str[pos+1]>>6)*9+((str[pos+1]-'0')&15)) << 4) |
```
Comment (achow101) [inline src/wallet/rpcdump.cpp]: Unrelated whitespace change.

File: src/wallet/rpcdump.cpp [hunk 2]
```diff
@@ -644,7 +644,7 @@ UniValue dumpwallet(const JSONRPCRequest& request)
     file << strprintf("#   mined on %s\n", EncodeDumpTime(chainActive.Tip()->GetBlockTime()));
     file << "\n";
 
-    // add the base58check encoded extended master if the wallet uses HD 
+    // add the base58check encoded extended master if the wallet uses HD
```
Comment (achow101) [inline src/wallet/rpcdump.cpp]: Unrelated whitespace change.

**Review by achow101 — COMMENTED** [review]
Concept ACK

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (MeshCollider) [reply src/wallet/rpcdump.cpp]: I mentioned that in the PR, is it an issue?

Comment (achow101) [reply src/wallet/rpcdump.cpp]: I isn't really a problem, but you should try to avoid making such changes. Atom can be set to not make any whitespace changes at all.

**Comment by MarcoFalke** [comment]
Mind to add to `dumpwallet` that just backing up the private masterkey is likely insufficient?

**Comment by MeshCollider** [comment]
Done :+1:

File: src/wallet/rpcdump.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/wallet/rpcdump.cpp]: It is a problem, as it often breaks backporting cherry-picks.

Comment (MeshCollider) [reply src/wallet/rpcdump.cpp]: Fixed.

**Comment by MeshCollider** [comment]
Rebased to fix .bitcoin directory travis failure

**Comment by achow101** [comment]
utACK 288dfaea9d987399b92514121f98a674e74d5580

File: src/wallet/rpcdump.cpp [hunk 3]
```diff
@@ -601,6 +601,7 @@ UniValue dumpwallet(const JSONRPCRequest& request)
         throw std::runtime_error(
             "dumpwallet \"filename\"\n"
             "\nDumps all wallet keys in a human-readable format.\n"
+            "Note that if your wallet contains keys which are not derived from your HD seed (e.g. imported keys), these are not covered by only backing up the seed itself, and must be backed up too.\n"
```
Comment (TheBlueMatt) [inline src/wallet/rpcdump.cpp]: This leaves me confused. The things I need to back up *are* in dumpwallet, just not the master key listed in dump wallet, right? So just the dumpwallet file is a sufficient backup (with rescan, including imported segwit addresses and p2sh redeemScripts?), but not only the master key?

**Review by TheBlueMatt — COMMENTED** [review]
I suppose for completeness importwallet may need the (rather astoundingly obvious) "Requires a new backup" note.

I guess we dont want to note anything about keypool topup leading to needing backups in things like "getnewaddress" (for non-HD wallets)? What about explicit ones like "keypoolrefill"? And things that always wipe keypool like encryptwallet? I guess probably not?

Similarly, I suppose we dont care about listing things like move/sendfrom which change wallet fileds for F$#King accounts?

File: src/wallet/rpcdump.cpp [hunk 4]
```diff
@@ -601,6 +601,7 @@ UniValue dumpwallet(const JSONRPCRequest& request)
         throw std::runtime_error(
             "dumpwallet \"filename\"\n"
             "\nDumps all wallet keys in a human-readable format.\n"
```
Comment (mess110) [inline src/wallet/rpcdump.cpp]: @TheBlueMatt would changing this string to something like `Dumps all wallet _derived_ keys in a human-readable format.` help cleanup the confusion?

File: src/wallet/rpcdump.cpp (hunk 3 above)
Comment (MeshCollider) [reply src/wallet/rpcdump.cpp]: I've changed it, lmk what you think now

**Comment by MeshCollider** [comment]
Rebased + hopefully clarified the text as per @TheBlueMatt comments

**Comment by TheBlueMatt** [comment]
Wait, now I'm even more confused. AFAICT, dumpwallet *does* dump all imported keys, not just "derived" keys, so no need to back those up. It does not, however, dump imported scripts, so witness addresses and multisig addresses will be missed, right?

**Comment by MeshCollider** [comment]
That's what I meant by `(e.g. ensure you back up the whole dumpfile)` (note that I didn't follow the suggested edit by mess110 because it's wrong, is that what you were looking at?). Are you suggesting mentioning that those imported scripts aren't backed up, or?

**Comment by TheBlueMatt** [comment]
Hmm, maybe I missed the parenthetical. Do we want to mention anything about the segwit addresses and imported scripts?

**Comment by MeshCollider** [comment]
The description does currently specifically refer to 'keys' not scripts or segwit addresses. But is there any reason scripts aren't also dumped too?

**Comment by laanwj** [comment]
> The description does currently specifically refer to 'keys' not scripts or segwit addresses. But is there any reason scripts aren't also dumped too?

No. just that (AFAIK) no one implemented that yet. Ideally a dumpwallet would not lose any information.

**Comment by TheBlueMatt** [comment]
I dont see why scripts shouldnt be dumped either. I'm not sure if its really in scope for this PR, but would be a welcome addition IMO.

**Comment by MeshCollider** [comment]
@laanwj @TheBlueMatt I'll look into it and try get a PR open for that soon too then

**Comment by MeshCollider** [comment]
Rebased

File: src/wallet/rpcdump.cpp [hunk 5]
```diff
@@ -601,6 +601,7 @@ UniValue dumpwallet(const JSONRPCRequest& request)
         throw std::runtime_error(
             "dumpwallet \"filename\"\n"
             "\nDumps all wallet keys in a human-readable format to a server-side file. This does not allow overwriting existing files.\n"
+            "Note that if your wallet contains keys which are not derived from your HD seed (e.g. imported keys), these are not covered by only backing up the seed itself, and must be backed up too (e.g. ensure you back up the whole dumpfile).\n"
```
Comment (TheBlueMatt) [inline src/wallet/rpcdump.cpp]: Note that you may want to wrap this line.

**Review by TheBlueMatt — COMMENTED** [review]
utACK a38bfbc51d0b98fffd8d79457f31c6fb196ff580. Would like to see the dumpwallet-includes-imported-scripts followup.

**Comment by MeshCollider** [comment]
I've wrapped the text, and added a note about scripts not being included, until I get the other PR up. Will remove that note in the other PR when it happens.

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -1110,7 +1110,7 @@ UniValue addmultisigaddress(const JSONRPCRequest& request)
     if (request.fHelp || request.params.size() < 2 || request.params.size() > 3)
     {
         std::string msg = "addmultisigaddress nrequired [\"key\",...] ( \"account\" )\n"
-            "\nAdd a nrequired-to-sign multisignature address to the wallet.\n"
+            "\nAdd a nrequired-to-sign multisignature address to the wallet. Requires a new wallet backup.\n"
```
Comment (JohnVonNeumann) [inline src/wallet/rpcwallet.cpp]: "nrequired-to-sign multisignature"

Only comment, I'm assuming this isn't a typo, and is a reference to the number of required signatures needed for the multisig address?

Comment (MeshCollider) [reply src/wallet/rpcwallet.cpp]: Correct :) But note that that text is not changed in this PR, it was there before

Comment (JohnVonNeumann) [reply src/wallet/rpcwallet.cpp]: Ok sweet, yeah I saw that it wasn't a change from this commit, but wasn't sure so figured I'd ask. Thanks! Is it typically frowned upon to comment on things not directly changed by the PR/Commit/Issue in question?

Comment (MeshCollider) [reply src/wallet/rpcwallet.cpp]: No, if something looks wrong then feel free to point it out :) I was just clarifying

Comment (JohnVonNeumann) [reply src/wallet/rpcwallet.cpp]: Awesome, thanks again.
